core/services/workflows/engine.go

```diff
@@ -795,6 +838,66 @@ func (e *Engine) deregisterTrigger(ctx context.Context, t *triggerCapability, tr
 	return nil
 }
 
+func (e *Engine) isWorkflowFullyProcessed(state store.WorkflowExecution) (bool, string, error) {
+	statuses := map[string]string{}
+	// we need to first propagate the status of the errored status if it exists...
+	err := e.workflow.walkDo(workflows.KeywordTrigger, func(s *step) error {
+		stateStep, ok := state.Steps[s.Ref]
+		if !ok {
+			return nil
+		}
+		statuses[s.Ref] = stateStep.Status
+		switch stateStep.Status {
+		case store.StatusErrored, store.StatusCompletedEarlyExit:
+			for _, sd := range s.Dependencies {
+				statuses[sd] = stateStep.Status
+			}
+		}
+		return nil
+	})
+	if err != nil {
+		return false, "", err
+	}
+
+	workflowProcessed := true
+	err = e.workflow.walkDo(workflows.KeywordTrigger, func(s *step) error {
+		if _, ok := state.Steps[s.Ref]; !ok {
+			if len(s.Dependencies) == 0 {
```

Comment:
Do we need to check dependencies here? At this point state should be a map where any missing element is pending, and as long as we have one pending item workflowProcessed should be marked as false.

Reply:
We do, because the missing element could be also a step that depends on an errored/exited_early step, and it is not on the DB state cause it never got executed, if that makes sense. Let me know if that's wrong, please.

Reply:
Ah I see: this should go away if we check against the statuses map I think?

So the algorithm here should be:
- walk the graph
- for each step, check the statuses entry
- if it's missing from statuses, it's pending because in all other cases we've added an entry (for errored, completed early exit, timeout)

Reply:
Thank you